source/extensions/filters/network/redis_proxy/router_impl.h

```diff
@@ -64,16 +66,21 @@ class Prefix : public Route {
 
 using PrefixSharedPtr = std::shared_ptr<Prefix>;
 
-class PrefixRoutes : public Router {
+class PrefixRoutes : public Router, public Logger::Loggable<Logger::Id::redis> {
 public:
   PrefixRoutes(const envoy::extensions::filters::network::redis_proxy::v3::RedisProxy::PrefixRoutes&
                    prefix_routes,
                Upstreams&& upstreams, Runtime::Loader& runtime);
 
   RouteSharedPtr upstreamPool(std::string& key) override;
 
-  void setReadFilterCallback(Network::ReadFilterCallbacks* callbacks) override {
-    callbacks_ = callbacks;
+  void initializeReadFilterCallbacks(Network::ReadFilterCallbacks* callbacks) override {
+    // Capture the stream info so that we can use it for formatting. This is needed to
+    // to avoid lifetime issues if the stream is destroyed before formatting is done.
+    stream_info_ = std::make_unique<StreamInfo::StreamInfoImpl>(
```

Comment:
We are facing few issues copying stream like this based on passed in callbacks. 

It seems the connectionProvider is getting mixed up with new connections. It happens before this change also (previously we were holding callbacks reference here).  Not exactly sure why though. May be holding callbacks reference here is not a good idea. We are experimenting by changing to pass callbacks in `makeRequest` and through the call chain to `upstreamPool` method.

Any other suggestions?